src/wp-includes/blocks.php

```diff
@@ -283,15 +283,38 @@ function get_block_metadata_i18n_schema() {
  * @return WP_Block_Type|false The registered block type on success, or false on failure.
  */
 function register_block_type_from_metadata( $file_or_folder, $args = array() ) {
-	$filename      = 'block.json';
-	$metadata_file = ( substr( $file_or_folder, -strlen( $filename ) ) !== $filename ) ?
-		trailingslashit( $file_or_folder ) . $filename :
+	/**
+	 * Get an array of metadata from a PHP file.
+	 * This improves performance for core blocks as it's only necessary to read a single PHP file
+	 * instead of reading a JSON file per-block, and then decoding from JSON to PHP.
+	 * Using a static var ensures that the metadata is read only once per request.
+	 */
+	static $core_blocks_meta;
+	if ( ! $core_blocks_meta ) {
+		$core_blocks_meta = include_once ABSPATH . WPINC . '/blocks/blocks-json.php';
+	}
+
+	$metadata_file = ( substr( $file_or_folder, -10 ) !== 'block.json' ) ? // 10 is the string-length of 'block.json'.
+		trailingslashit( $file_or_folder ) . 'block.json' :
 		$file_or_folder;
 	if ( ! file_exists( $metadata_file ) ) {
 		return false;
 	}
 
-	$metadata = wp_json_file_decode( $metadata_file, array( 'associative' => true ) );
+	// Try to get metadata from the static cache for core blocks.
+	$metadata = false;
+	if ( str_starts_with( $file_or_folder, ABSPATH . WPINC ) ) {
+		$core_block_name = str_replace( ABSPATH . WPINC . '/blocks/', '', $file_or_folder );
```

Comment:
Have you considered using the same paths passed to the `register_block_type` as keys in the generated file? This way, you could do a simple check:

```php
if ( ! empty( $core_blocks_meta[ $file_or_folder ] ) ) {
    $metadata = $core_blocks_meta[ $file_or_folder ];
}
```

Reply:
Aren't those paths absolute - which means they'd change from site to site? 🤔

Reply:
I was thinking you could set a variable to use in keys:

```php
$dir = ABSPATH . WPINC . '/blocks/';
```

Maybe, it's too much work in a bit different place 😄